src/utils/sample/uniform.rs

```diff
@@ -0,0 +1,256 @@
+use rand::{rngs::ThreadRng, RngCore};
+
+#[allow(dead_code)]
+static mut RNG: Option<ThreadRng> = None;
+
+/// Returns a pointer to a static [`ThreadRng`] to sample uniform at random.
+/// If the Random Generator was not used/ instantiated before,
+/// it initializes the static variable.
+///
+/// **WARNING:** This function is strictly for single-threaded use!
+/// Otherwise access needs to be synchronized.
+///
+/// # Examples
+/// ```compile_fail
+/// use qfall_math::utils::sample::get_rng;
+///
+/// let rng = get_rng();
+/// ```
+#[allow(dead_code)]
+fn get_rng() -> &'static mut ThreadRng {
+    if unsafe { RNG.is_none() } {
+        unsafe { RNG = Some(ThreadRng::default()) };
+    }
+
+    unsafe { RNG.as_mut().unwrap() }
+}
+
+/// Computes a uniform at random chosen [`Z`] sample in `[0, interval_size)`.
+///
+/// Parameters:
+/// - `interval_size`: specifies the size of the interval
+/// over which the samples are drawn
+///
+/// Returns a uniform at random chosen [`Z`] instance in `[0, interval_size)`.
+///
+/// # Examples
+/// ```compile_fail
+/// use qfall_math::utils::sample::sample_uniform_rejection;
+/// use qfall_math::integer::Z;
+/// let interval_size = Z::from(20);
+///
+/// let sample = sample_uniform_rejection(&interval_size);
+///
+/// assert!(0 <= sample);
+/// assert!(sample < interval_size);
+/// ```
+///
+/// # Errors and Failures
+/// - Returns a [`MathError`] of type [`InvalidInterval`](MathError::InvalidInterval)
+/// if the interval is chosen smaller than or equal to `1`.
+#[allow(dead_code)]
+pub(crate) fn sample_uniform_rejection(interval_size: &Z) -> Result<Z, MathError> {
+    if interval_size <= &Z::ONE {
+        return Err(MathError::InvalidInterval(format!(
+            "An invalid interval size {} was provided.",
+            interval_size
+        )));
+    }
+
+    let bit_size = interval_size.bits() as usize;
+    let bit_vector = sample_bits_uniform(bit_size);
+
+    let random = Z::from(&bit_vector);
+    if &random < interval_size {
+        Ok(random)
+    } else {
+        sample_uniform_rejection(interval_size)
+    }
+}
+
+/// Computes `nr_bits` many uniform at random chosen bits.
+///
+/// Parameters:
+/// - `nr_bits`: specifies the number of bits to be chosen uniform at random
+///
+/// Returns a [`Vec<u8>`] including `nr_bits` many uniform at random chosen bits,
+/// filled with `0` bits if needed for the last byte resp. [`u8`].
+///
+/// # Examples
+/// ```compile_fail
+/// use qfall_math::utils::sample::sample_bits_uniform;
+/// let nr_bits = 6;
+///
+/// let byte_vector = sample_bits_uniform(nr_bits);
+///
+/// assert_eq!(0, byte_vector[6]);
+/// assert_eq!(0, byte_vector[7]);
+/// assert_eq!(2, byte_vector.len());
+/// ```
+#[allow(dead_code)]
+fn sample_bits_uniform(nr_bits: usize) -> Vec<u8> {
+    let rng = get_rng();
+
+    // sample ⌈ nr_bits / 8 ⌉ bytes
+    let mut byte_vector;
+    if nr_bits % 8 == 0 {
+        byte_vector = vec![0_u8; nr_bits / 8];
```

Comment:
why use 0_u8 here?
If it is needed, use 0u8 like in  line 118

Reply:
✓